### ISSUE TITLE
Kubernetes master unavalability triggers cluster restart

### ISSUE THREAD

**Issue by bdubertret** [rabbitmq/cluster-operator#709]
## Describe the bug
Whenever the Kubernetes master is unavailable, the cluster-operator crashes as described in #69, because it can't access the kubernetes API. When the kubernetes master comes back, it restarts the cluster-operator, which in turn triggers a cluster restart on managed clusters.

## To Reproduce
1. Deploy cluster-operator and RabbitMQ cluster on kubernetes
2. Make kubernetes master unavailable (e.g., trigger a master upgrade)
3. Wait for kubernetes master to come back
4. Managed clusters get restarted

**Expected behavior**
Managed clusters should not be restarted

## Version and environment information
- RabbitMQ: 3.8.12
- RabbitMQ Cluster Operator: 1.5.0
- Kubernetes: 1.17
- Cloud provider or hardware configuration: GKE

**Comment by MirahImage** [comment]
Hi @bdubertret, I was unable to reproduce your issue on GKE using either v1.5.0 or the current version of the cluster operator (v1.7.0). 

While Kubernetes API downtime due to a master upgrade did cause the cluster-operator pod to crash and be restarted several times, leading to a crashLoopBackOff lasting for several minutes, this did not result in a restart of the RabbitMQ cluster. 

Separately, I was able to cause the RabbitMQ cluster to be restarted by upgrading the Kubernetes nodes, which caused the nodes to be drained, forcing the pods to be moved to new nodes, but this was unrelated to the API downtime or operator being restarted.

**Comment by bdubertret** [comment]
Hi @MirahImage, I tried with the current version of the cluster operator (v1.7.0), and I managed to reproduce the problem, on our test GKE cluster.

Here is the RabbitMQ cluster definition I used : 

```
apiVersion: rabbitmq.com/v1beta1
kind: RabbitmqCluster
metadata:
  name: rabbitmq
spec:
  replicas: 3
  resources:
    limits:
      cpu: 256m
      memory: 1Gi
    requests:
      cpu: 256m
      memory: 1Gi
  service:
    type: NodePort
```

I applied the cluster-operator.yml from https://github.com/rabbitmq/cluster-operator/releases/download/v1.7.0/cluster-operator.yml

My current wild guess is that this cluster definition lacks something, left for cluster-operator to define at runtime. When the k8s master restarts the cluster-operator pod, it in turn leads to a reconciliation process that modifies the StatefulSets, leading the k8s master to restart the RabbitMQ pods.

During this test, I created the GKE cluster, deployed the cluster-opertor.yml, and the above RabbitmqCluster definition, waited for it to settle with all three rabbitmq pods marked as ready, and then upgraded the GKE master version, from 1.17 to 1.18. 

The timeline of the GKE master upgrade is given by the `gcloud container operations list` as : 

```
NAME                              TYPE                LOCATION        TARGET                         STATUS_MESSAGE  STATUS  START_TIME                      END_TIME
operation-xxx-xxx  UPGRADE_MASTER      europe-west1-b  xxx                            DONE    2021-06-02T09:44:56.303826154Z  2021-06-02T09:49:19.222651274Z
```
The cluster-operator pod crashed during the GKE master upgrade, and gets restarted at Wed, 02 Jun 2021 09:51:50 +0000.

Here is the log of this cluster-operator pod : 
[rabbitmq-cluster-operator.log](https://github.com/rabbitmq/cluster-operator/files/6584990/rabbitmq-cluster-operator.log)

We can see it fail to access the Kubernetes API beginning at 2021-06-02T09:45:21.059692921Z, and then crash. It is restarted at 2021-06-02T09:51:56.060471221Z and then begins reconciling at 2021-06-02T09:52:11.031171605Z. This reconciliation process in turn leads to the Statefulsets to get restarted : 

```
$ kubectl get pod rabbitmq-server-{0,1,2} -o jsonpath='{.items[*].status.startTime}'                                                                                                                     2021-06-02T09:56:13Z 2021-06-02T09:54:27Z 2021-06-02T09:52:47Z
```

I was tempted to use the new settings introduced by #111 but this does not seem to be related, as these timeouts are expressed in seconds while this k8s master unavailability problem lasts for minutes. And I guess this is not an easy fix, because who knows, at the cluster-operator level, what causes this k8s API unavailability. 

But in the meantime, I'd like to find a solution so that this cluster operator crash does not lead to an unwanted rabbitmq cluster restart. In the GKE context, we can't really decide when GKE master gets updated, so this causes trouble in production systems.

**Comment by MirahImage** [comment]
In the case of an upgrade, the k8s API becomes unavailable (for example, due to running in a single-master environment), the cluster operator will crash, but that on its own seems to be insufficient to cause a RabbitMQ rolling upgrade, as I verified by performing repeated upgrades on GKE yesterday. 

From the logs, it's clear that the operator did indeed attempt to reconcile changes to the StatefulSet after being restarted, but the exact reason for this is unclear. In order to help us figure this out, could you provide the output of `kubectl get pod <rabbit-pod> -o yaml` from before and after the upgrade (or really, just the diff) so we can figure out what's causing the problem? If there's no diff, perhaps there's a diff in the statefulset itself?

Thanks for helping us track this down.

**Comment by bdubertret** [comment]
I did a new test on the same GKE cluster and using the same items, the reconciliation problem did not happen after a new GKE master upgrade. So I went back to set up a fresh GKE cluster, deploy cluster-operator.yml (1.7.0) and the above mentioned RabbitmqCluster definition, triggered a GKE master upgrade from 1.17 to 1.18, and met the reconciliation problem. It seems to only happen once per rabbitmq cluster deployment.

As you asked, I took a snapshot of the RabbitmqCluster CRD, Statefulset, and pods, using `kubectl get [...] -o yaml` before the GKE master upgrade, and after the upgrade, and ran a diff. It seems that the most relevant part is a change in the Statefulset definition, namely : 

```
+      topologySpreadConstraints:
+      - labelSelector:
+          matchLabels:
+            app.kubernetes.io/name: rabbitmq
+        maxSkew: 1
+        topologyKey: topology.kubernetes.io/zone
+        whenUnsatisfiable: ScheduleAnyway
```

So, it seems that this part of the Statefulset isn't set at first, but set after the cluster operator restarts following the GKE master upgrade. As per the kubernetes documentation, `topologySpreadConstraints` is an alpha feature in kubernetes 1.17 and a beta feature in 1.18, which might explain why it is not in the first version while it is in the second. This change in the StatefulSet seems to be enough for k8s to trigger a restart of the rabbitmq pods.

So, it seems that this problem arises because the Kubernetes API does not consider the  `topologySpreadConstraints` in version 1.17 but does in 1.18, which leads to the Statefulset pods being restarted, and GKE upgrades its master without prior notice. It is not a problem with the rabbitmq cluster operator itself : this particular part of the Statefulset is set in https://github.com/rabbitmq/cluster-operator/blob/v1.7.0/internal/resource/statefulset.go#L518 only to be ignored in k8s 1.17. Using the production-ready settings in https://github.com/rabbitmq/cluster-operator/blob/v1.7.0/docs/examples/production-ready/rabbitmq.yaml wouldn't have helped either in this case, for the same reason (ignored in 1.17, used in 1.18).

Thanks for your time taken to look at this issue, I will close this issue now.
